sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java

```diff
@@ -74,25 +81,73 @@ public void roslyn_metrics_are_imported() {
+    var inputFile = addTestFileToContext("Cases.razor");
+    var noSonarFilter = mock(NoSonarFilter.class);
+    var fileLinesContext = mock(FileLinesContext.class);
+    var fileLinesContextFactory = mock(FileLinesContextFactory.class);
+    when(fileLinesContextFactory.createFor(any(InputFile.class))).thenReturn(fileLinesContext);
+
+    new MetricsImporter(sensorContext, fileLinesContextFactory, noSonarFilter, RazorImporterTestBase::fileName).accept(PROTOBUF_4_10_FILE.toPath());
+
+    var measures = sensorContext.measures(inputFile.key());
+    assertThat(measures).hasSize(7);
+
+    assertThat(measures).extracting("metric", "value")
+      .containsOnly(
+        Tuple.tuple(CoreMetrics.COMPLEXITY, 5),
+        Tuple.tuple(CoreMetrics.FUNCTIONS, 3),
+        Tuple.tuple(CoreMetrics.COMMENT_LINES, 0),
+        Tuple.tuple(CoreMetrics.COGNITIVE_COMPLEXITY, 1),
+        Tuple.tuple(CoreMetrics.CLASSES, 0),
+        Tuple.tuple(CoreMetrics.NCLOC, 14),
+        Tuple.tuple(CoreMetrics.STATEMENTS, 3));
+
+    verify(noSonarFilter).noSonarInFile(inputFile, Collections.emptySet());
+
+    verifyMetrics(fileLinesContext, CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 8, 23, 24);
+    verifyMetrics(fileLinesContext, CoreMetrics.NCLOC_DATA_KEY, 1, 3, 5, 8, 9, 13, 16, 18, 19, 21, 22, 23, 24, 25);
+
+    assertThat(logTester.logs(Level.DEBUG)).isEmpty();
+  }
+
+  @Test
+  public void roslyn_metrics_out_of_range_with_4_10_debug_enabled() throws FileNotFoundException {
+    addTestFileToContext("_Imports.razor");
+    var fileLinesContextFactory = mock(FileLinesContextFactory.class);
+    when(fileLinesContextFactory.createFor(any(InputFile.class))).thenReturn(mock(FileLinesContext.class));
+
+    new MetricsImporter(sensorContext, fileLinesContextFactory, mock(NoSonarFilter.class), RazorImporterTestBase::fileName).accept(PROTOBUF_4_10_FILE.toPath());
+
+    assertThat(logTester.logs(Level.DEBUG)).containsExactly("The code line number was out of the range. File _Imports.razor, Line 4");
+  }
+
+  @Test
+  public void roslyn_metrics_out_of_range_with_4_10_debug_disabled() throws FileNotFoundException {
+    logTester.setLevel(Level.INFO);
+    addTestFileToContext("_Imports.razor");
+    var fileLinesContextFactory = mock(FileLinesContextFactory.class);
+    when(fileLinesContextFactory.createFor(any(InputFile.class))).thenReturn(mock(FileLinesContext.class));
+
+    new MetricsImporter(sensorContext, fileLinesContextFactory, mock(NoSonarFilter.class), RazorImporterTestBase::fileName).accept(PROTOBUF_4_10_FILE.toPath());
+
+    assertThat(logTester.logs(Level.DEBUG)).isEmpty();
+  }
+
+  private void verifyMetrics(FileLinesContext context, String key, int... values) {
+    var groups = Arrays.stream(values)
+      .boxed()
+      .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+    for (int groupKey : groups.keySet()){
+      verify(context, times(groups.get(groupKey).intValue())).setIntValue(key, groupKey, 1);
```

Comment:
I find it a bit hard to understand what you are doing here.
Why do you group the lines and then do `times`?
From debugging they are always the unique.
How could it be that a group has more than one occurrence of each line?

Do you want to make sure exactly this? That every line appears ONLY once?

Reply:
When multiple files with test cases are added to the context, we can have lines reported multiple times. This was the behavior before my changes when the test base class added files to the context.

Strictly speaking, this is now used only to check that each element is added only once, and the query can be simplified. I wanted to have this method able to check duplicates, so we can easily add test cases in the future.

Reply:
That makes sense, thanks!